dotnet/src/Experimental/Process.Core/ProcessBuilder.cs

```diff
@@ -0,0 +1,164 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.SemanticKernel;
+
+/// <summary>
+/// Provides functionality for incrementally defining a process.
+/// </summary>
+public class ProcessBuilder : ProcessStepBuilder
+{
+    private readonly List<ProcessStepBuilder> _steps;
+    private readonly List<ProcessStepBuilder> _entrySteps;
+    private readonly Dictionary<string, ProcessStepBuilder> _stepsMap;
+
+    /// <summary>
+    /// Used to resolve the target function and parameter for a given optional function name and parameter name.
+    /// This is used to simplify the process of creating a <see cref="KernelProcessFunctionTarget"/> by making it possible
+    /// to infer the function and/or parameter names from the function metadata if only one option exists.
+    /// </summary>
+    /// <param name="functionName">The name of the function. May be null if only one function exists on the step.</param>
+    /// <param name="parameterName">The name of the parameter. May be null if only one parameter exists on the function.</param>
+    /// <returns>A valid instance of <see cref="KernelProcessFunctionTarget"/> for this step.</returns>
+    /// <exception cref="InvalidOperationException"></exception>
+    internal override KernelProcessFunctionTarget ResolveFunctionTarget(string? functionName, string? parameterName)
+    {
+        // Try to resolve the function target on each of the registered entry points.
+        var targets = new List<KernelProcessFunctionTarget>();
+        foreach (var step in this._entrySteps)
+        {
+            try
+            {
+                targets.Add(step.ResolveFunctionTarget(functionName, parameterName));
+            }
+            catch (InvalidOperationException)
+            {
+                // If the function is not found on the source step, then we can ignore it.
+            }
+        }
+
+        // If no targets were found or if multiple targets were found, throw an exception.
+        if (targets.Count == 0)
+        {
+            throw new InvalidOperationException($"No targets found for the specified function and parameter '{functionName}.{parameterName}'.");
```

Comment:
(This comment applies to all instances throw `InvalidOperationException` as well as the `catch` on line #37 of this file)

I might consider using either `KernelException` or introducing a `ProcessException` that extends `KernelException` (I guess that's allowed now - see Vector Store framework).

Over using system exceptions may not be ideal for both internal exception management as well as that of the user of the _Process Framework_.